### PR TITLE
SNOW-1675257 AST encoding for Stored Procedures

### PR DESCRIPTION
**Pull request by sfc-gh-vbudati**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-1675257

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

Added AST encoding for Stored Procedures.

- This involved making a few changes in the local testing API to help with testing the sproc changes.
- The only relevant test here should be `sproc.test`, all of the other tests are updated because of `--update-expectations`.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/ast_utils.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 import os
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/_internal/ast_utils.py:9]: can you update with latest [ls-SNOW-1491199-merge-phase0-server-side](https://github.com/snowflakedb/snowpark-python/tree/ls-SNOW-1491199-merge-phase0-server-side) changes?

File: src/snowflake/snowpark/_internal/ast_utils.py [hunk 2]
```diff
@@ -379,12 +380,25 @@ def build_udaf_apply(
...
 
 
+def build_sproc_apply(
+    ast: proto.Expr,
+    sproc_id: int,
+    statement_params: Optional[Dict[str, str]] = None,
+    *args: Tuple[Union[proto.Expr, Any]],
+    **kwargs,
+) -> None:
+    """Builds call to stored procedure into ast as a Snowpark IR expression."""
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/_internal/ast_utils.py:396]: nit: Encodes call to stored procedure object.

File: src/snowflake/snowpark/_internal/ast_utils.py [hunk 3]
```diff
@@ -1179,3 +1193,85 @@ def build_udtf(
...
+    comment: Optional[str] = None,
+    statement_params: Optional[Dict[str, str]] = None,
+    execute_as: typing.Literal["caller", "owner"] = "owner",
+    source_code_display: bool = True,
+    is_permanent: bool = False,
+    session=None,
+    **kwargs,
+):
+    """Helper function to encode stored procedure parameters (used in both regular and mock StoredProcedureRegistration)."""
+    # TODO: to unparse/reference callables client-side - track them in ast_batch.
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/_internal/ast_utils.py:1222]: can we resolve this TODO? it should be fairly easy :)
Afaik this here should take core of it
```
 build_proto_from_callable(
        ast.func, func, session._ast_batch if session is not None else None
    )
```

Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/_internal/ast_utils.py:1222]: Ooh sorry this was an artifact from copying over the code in `build_udf`. Should I get rid of the TODO over there as well?

File: tests/ast/data/Session.call.test [hunk 4]
```diff
@@ -12,4 +12,4 @@ df2 = session.call("my_sproc")
```
Comment (sfc-gh-lspiegelberg) [inline tests/ast/data/Session.call.test:12]: with sprocs working now, can we fix up this test so "my_sproc" gets properly registered here first? (You can do so either in test setup, or here)

File: src/snowflake/snowpark/functions.py [hunk 5]
```diff
@@ -9538,9 +9537,6 @@ def sproc(
         2
     """
 
-    if _emit_ast:
-        raise NotImplementedError("TODO 1675257: support sprocs for AST.")
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/functions.py:9542]: Did you search for `"1675257"`? No more TODO referencing this ticket in the source code?

File: src/snowflake/snowpark/mock/_plan.py [hunk 6]
```diff
@@ -595,7 +595,7 @@ def handle_udaf_expression(
             )
             if coerce_result is None:
                 raise SnowparkLocalTestingException(
-                    f"UDF received input type {column_data.sf_type.datatype} for column {child.name}, but expected input type of {expected_type}"
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/mock/_plan.py:598]: good catch!

File: src/snowflake/snowpark/_internal/ast_utils.py (hunk 3 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/_internal/ast_utils.py:1222]: Discussed on slack - removed both TODOs here. Verified that the `id` field of `SpCallable` is set.

File: src/snowflake/snowpark/functions.py (hunk 5 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/functions.py:9542]: Yup! I haven't seen anything else referencing this ticket.

File: src/snowflake/snowpark/stored_procedure.py [hunk 7]
```diff
@@ -79,9 +88,15 @@ def __init__(
         self._execute_as = execute_as
         self._anonymous_sp_sql = anonymous_sp_sql
         self._is_return_table = isinstance(return_type, StructType)
-
         self._packages = packages
 
+        # If None, no ast will be emitted. Else, passed whenever sproc is invoked.
+        self._ast = _ast
+        self._ast_id = _ast_id
+        self._stmt = (
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/stored_procedure.py:96]: nit: you can move the comment to the line above and do
```
self._stmt = _stmt
```

File: src/snowflake/snowpark/functions.py (hunk 5 above)
Comment (sfc-gh-lspiegelberg) [reply src/snowflake/snowpark/functions.py:9542]: ok, perfect!

File: src/snowflake/snowpark/mock/_stored_procedure.py [hunk 8]
```diff
@@ -141,6 +168,14 @@ def __call__(
         ) and not isinstance(result, DataFrame):
             result = json.dumps(result, indent=2, cls=CUSTOM_JSON_ENCODER)
 
+        if self._is_return_table:
+            # If the result is a Column or DataFrame object, the expression `eval` is performed in a later operation
```
[the remark this answers is not in the document]
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/mock/_stored_procedure.py:172]: Oh sorry, I meant that as in "the evaluation of the expression" not that `eval` is an expression; I can see how that is misread I'll reword that. 😅 But thanks for clarifying!! 😄

File: tests/ast/data/Session.call.test (hunk 4 above)
Comment (sfc-gh-vbudati) [reply tests/ast/data/Session.call.test:25]: done!

File: src/snowflake/snowpark/stored_procedure.py (hunk 7 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/stored_procedure.py]: done!

File: src/snowflake/snowpark/_internal/ast_utils.py (hunk 2 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/_internal/ast_utils.py]: done!

**Review by sfc-gh-lspiegelberg — APPROVED** [review]
lgtm

**Review by sfc-gh-lspiegelberg — CHANGES_REQUESTED** [review]
please fix merge gates :)

**Review by sfc-gh-lspiegelberg — APPROVED** [review]

**Pull request merged**